txpipe/source_selector.py

```diff
@@ -673,9 +469,252 @@ def select_2d(self, data, calling_from_select=False):
+        pz_data = {}
+        variants = ['00/', '1p/', '2p/', '1m/', '2m/']
+        for v in variants:
+            zz = shear_data[f'{v}mean_z']
+
+            pz_data_v = np.zeros(len(zz), dtype=int) -1
+            for zi in range(len(self.config['source_zbin_edges'])-1):
+                mask_zbin = (zz>=self.config['source_zbin_edges'][zi]) & (zz<self.config['source_zbin_edges'][zi+1])
+                pz_data_v[mask_zbin] = zi
+
+            pz_data[f'{v}zbin'] = pz_data_v
+
+        return pz_data
+
+
+    def setup_output(self):
+        outfile = super().setup_output()
+        n = outfile['tomography/source_bin'].size
+        nbin_source = outfile['tomography/source_counts'].size
+        group = outfile.create_group('response')
+        group.create_dataset('R', (nbin_source,2,2), dtype='f')
+        group.create_dataset('R_2d', (2,2), dtype='f')
+        return outfile
+
+    def compute_output_stats(self, calculator, mean, variance):
+        R, N = calculator.collect(self.comm, allgather=True)
+        calibrator = MetaDetectCalibrator(R, mean, mu_is_calibrated=False)
+        mean_e = calibrator.mu
+
+        P = np.diag(np.linalg.inv(R @ R))
+        # Apply to the variances to get sigma_e
+        sigma_e = np.sqrt(0.5 * P @ variance)
+
+        return BinStats(N, N, mean_e, sigma_e, calibrator)
+
+
+class TXSourceSelectorLensfit(TXSourceSelectorBase):
+    name = "TXSourceSelectorLensfit"
+    def data_iterator(self):
+        chunk_rows = self.config['chunk_rows']
+        bands = self.config['bands']
+        shear_cols = ['psf_T_mean', 'weight', 'flags', 'T', 's2n', 'g1', 'g2','weight','m']
+        shear_cols += band_variants(bands, 'mag', 'mag_err', shear_catalog_type='lensfit')
+        if self.config['input_pz']:
+            shear_cols += ['mean_z']
+        elif self.config['true_z']:
+            shear_cols += ['redshift_true']
+        return self.iterate_hdf('shear_catalog', 'shear', shear_cols, chunk_rows)
+
+    def setup_response_calculators(self, nbin_source):
+        calculators = [LensfitCalculator(self.select,self.config['input_m_is_weighted']) for i in range(nbin_source)]
+        calculators.append(LensfitCalculator(self.select_2d,self.config['input_m_is_weighted']))
+        return calculators
+
+    def setup_output(self):
+        outfile = super().setup_output()
+        n = outfile['tomography/source_bin'].size
+        nbin_source = outfile['tomography/source_counts'].size
+        group = outfile.create_group('response')
+        group.create_dataset('K', (nbin_source,), dtype='f')
+        group.create_dataset('C', (nbin_source,2), dtype='f')
+        group.create_dataset('K_2d', (1,), dtype='f')
+        group.create_dataset('C_2d', (2), dtype='f')
+        return outfile
+
+
+    def compute_output_stats(self, calculator, mean, variance):
+        K, C, N = calculator.collect(self.comm, allgather=True)
+        calibrator = LensfitCalibrator(K, C)
+        mean_e = C
```

Comment:
I didn't check how mean_e and C are used later, but would something like mean_e = C.copy() be less dangerous?

Reply:
good idea - done.